### ISSUE TITLE
Is there a way to share notebooks via JupyterHub?

### ISSUE THREAD

**Issue by zer0n** [jupyterhub/jupyterhub#394]
For example: make a notebook public to the team or multiple users

**Comment by minrk** [comment]
Not at this point, no. We are working on that.

**Comment by zer0n** [comment]
Great. Thanks! I look forward to that feature.

**Comment by nehalecky** [comment]
Seems related: https://github.com/jupyter/jupyterhub/issues/424. Perhaps helpful for @zer0n.

**Comment by rawrgulmuffins** [comment]
Would it be hard to add a permanent link feature to JupyterHub?

**Comment by minrk** [comment]
@rawrgulmuffins there needs to be a publication mechanism so that notebooks leave the single-user server and go to a place where they are accessible to other users. The 'permalink' has to be handled by a different server than the single-user notebook server.

**Comment by rawrgulmuffins** [comment]
@minrk would that be a small enough addition that a (non-core member) could include it in a week?

**Comment by tpacker** [comment]
I would love the ability to share notebooks with a URL

**Comment by hnykda** [comment]
This is probably the only feature which is stopping us from using JupyterHub.

**Comment by desilinguist** [comment]
Same here. We would really like to use JupyterHub but this is the blocker for now.

**Comment by minrk** [comment]
@rawrgulmuffins sorry for not responding. It depends on your needs how simple is appropriate. @dsblank has implemented a sharing service for [his deployment](https://github.com/dsblank/jupyter.brynmawr). I hope to have an official implementation by the end of the year (I had hoped by the end of this summer, but it's been a crazy busy summer).

**Comment by dsblank** [comment]
Specifically this file: 
https://github.com/dsblank/jupyter.brynmawr/blob/master/jupyterhub/publichandler.py

Combined with this nbextension to copy files to the public_html folder:
https://github.com/Calysto/notebook-extensions/tree/master/calysto/publish

These are hacky versions, quickly written to do the bare necessities, with hardcoded things. 

But I agree: our jupyterhub deployment needs such a service to be complete.

**Comment by rawrgulmuffins** [comment]
I think I'm going to attempt to tackle this next week then. If nothing else I can add a set of lessons learned to this issue.

**Comment by nvenayak** [comment]
This could be really useful in my use case: gathering and sharing of scientific data between a lab group and then making notebooks public upon the work being accepted to a journal.

I was considering hacking together tmpnb with hub. Anything to be made public would be copied to the tmpnb server, which could then be handled with its spawner which can also deal with culling unsused instances.

Maybe it's more elegant to do everything within hub and combine the spawners?

**Comment by tamasgal** [comment]
Any news on that? We are currently also looking for a solution to make publicly viewable notebooks...

**Comment by djvulee** [comment]
Yes, I also want this feature, is there any news?

**Comment by mraky** [comment]
yesyesyes ;)  this feature is very important...  users should have posibility to share notebook - but in controled manner... i.e accessible under name

anyone found any workaround for this ?

**Comment by m4ce** [comment]
I would also love to see this feature implemented. We have resorted to commit the notebooks into bitbucket and use the [iPython notebook plugin](https://marketplace.atlassian.com/plugins/com.atlassian.bitbucket.server.plugin.bitbucket-ipynb-plugin/server/overview) to render them.

👍

**Comment by willingc** [comment]
Hi @mc4ce and others,

You may be interested in following the work being done in the HubShare project https://github.com/jupyterhub/hubshare

**Comment by SimplicityGuy** [comment]
@willingc hubshare looks promising... however the docs clearly state that it doesn't work yet.

Is hubshare the way Jupyter plans to support sharing notebooks? What's the timeline for that project?

**Comment by minrk** [comment]
Yes, HubShare is one planned way to share notebooks with JupyterHub (not counting real-time collaboration). There are already many deployments that have implemented sharing in one way or another, such as mounting a filesystem with its own sharing semantics (e.g. ownCloud) or extensions for publishing to S3. I hope to have a working version of HubShare by the end of this Spring.

**Comment by jai11** [comment]
Any beta version for notebook sharing ? or code ?

**Comment by joaoamaro70** [comment]
I also want this feature, is there any news?

**Comment by FalseProtagonist** [comment]
This would be nice!

**Comment by pranayhasan** [comment]
Hubshare looks exciting. It's been about an year and a half. Does it take more time to get the Jupyter team to actively work on it?

**Comment by willingc** [comment]
Folks may also be interested in following the progress of the https://github.com/jupyterlab/jupyterlab-google-drive project too.

**Comment by mraky** [comment]
i know issues on github are not for voting, still i will add + for this enhancement...  What i am missing in discussion here is, sharing of jupyter notebook is not only about sharing. It is necessary to have( as part of it ) some kind of scheduler. I see some are proposing google drive. I am not sure about your situation, but on my project i am not allowed to use any public resource.....

i dream about button, which will allow me schedule execution of notebook, and then generate link, there i will specify access rights to notebook ... i.e view only , hide code, full access etc...

**Comment by dsblank** [comment]
I've updated our jupyterhub PublicHandler to be a JupyterHub Service:

https://github.com/BrynMawrCollege/jupyterhub 

This will allow jupyterhub to render notebooks (like nbviewer) that are found in a user's public_html folder. There is also an associated button that copies notebooks to the public_html folder.

There are still some things hardcoded here, but might serve as a useful starting point. Also, Pull Requests accepted :)

**Comment by willingc** [comment]
@dsblank Thanks for the update. Would you mind opening an issue on jupyterhub/hubshare sharing this work as an example service? :cookie:

**Comment by dsblank** [comment]
Absolutely! Thanks for the suggestion, and cookie :)

**Comment by dsblank** [comment]
Done at https://github.com/jupyterhub/hubshare/issues/15

**Comment by willingc** [comment]
Thanks @dsblank. I appreciate it.

**Comment by vicpara** [comment]
Any updates on this ? This is really an important feature and I'd like to +1 .

**Comment by willingc** [comment]
Hi @vicpara, 

Following the JupyterHub and Hubshare releases as well as viewing the Jupyter mailing list, would be the places to keep up to date. We agree it's an important feature :sunny:

**Comment by radudragusin** [comment]
In my company we started using [nbgallery](https://github.com/nbgallery/nbgallery) and are extremely happy with it. It has great integration with JupyteHub itself, through [nbgallery-extensions](https://github.com/nbgallery/nbgallery-extensions) and it has support for collaboration, with change request workflow, recommendation engine, ownership groups and a lot of additional stuff. It is quite a mature project and you can see it is used actively in an organization setting by the developers.

It is a Ruby on Rails project, so it took a bit of effort to have it deployed, but it really was a game-changer in enabling us to collaborate and build on each other's notebooks.

**Comment by elgalu** [comment]
Thanks for sharing @radudragusin I'll definitely take a look.

**Comment by hemantasingh** [comment]
Hi All,

I did it in very simple way:
   1. Create a public folder
   2. Place all the reviewed file in this folder.
   3. Create a UI nbextension with a modal popup which list all the files
   4. User select [input box] the particular file.
   5. Then copy the file using Server extension to home/dashboard page.

Just copy those file . that's all

**Comment by davidedelvento** [comment]
If I am not misunderstanding what you, @hemantasingh , wrote amd dod, I think your use case is very different from what other people here are seeking. 
The problem most people have (certainly I) is not "giving the copy of a notebook" to people (email will work for that too), the problem is collaboratively edit the **same** notebook, a-la Google Docs.

**Comment by sbailey-auro** [comment]
Here is a mockup
![shareable-anon 1](https://user-images.githubusercontent.com/30668418/39888890-1404f3a6-544b-11e8-8679-d8d1c442e311.png)

**Comment by gsemet** [comment]
the complex part is to centralize who has access to what notebook.

**Comment by sbailey-auro** [comment]
Doesn't the JupyterHub Prox do exactly that (provide a central choke point for authentication) already evidence of the option to bind JupyterHub access to Single Sign On (SSO) for example Okta?

![permission-examples](https://user-images.githubusercontent.com/30668418/39889707-79855dfe-544d-11e8-975b-0a6435da2ba4.png)

**Comment by gsemet** [comment]
yes you can say if a user can start/access to a jupyter server. But what if, when you share a notebook, the user shutdown its instance? how the reader jupyter server can fetch the notebook from shared instance? By downloading it? 
ideally yes, the feature like google docs would be great!

**Comment by sbailey-auro** [comment]
By the way, as a creepy work around, if you run a terminal on your JupyterHub spawned single-user notebook server, you can execute `jupyter notebook list`. From the output of that command, you'll get a token that you can append to the URL  that you use to access your Jupyter Notebook server (i.e. the JupyterHub proxy URL), and you can share that full URL out to others so they may have access to your running notebooks.

**Comment by sbailey-auro** [comment]
@gsemet  To do it "right" Jupyter itself should first be extended for stateful multi-user collaborative access which includes a change log (just like Google Suite).   Then all the permissions and sharing stuff would be added to JupyterHub....sounds like a nice Masters CS Project in online-collaborative systems..with Google Suite as the "model" covering most of the functionality and maybe a couple of interesting things relating to "execution cells" rather than just documents making it novel enough for a Master's CS Project.    😃

**Comment by CordThomas** [comment]
Can I ask what the latest is on how best to enable notebook sharing using JupyterHub?   This is one of our most useful uses of the hub.

**Comment by richyanicky** [comment]
Can you backend the storage of the notebooks to git once they are published? Then the user could send a URL to the git reference of the notebook. We have a need where the recipient of the notebook will need to run and maybe modify it. If they modify it they can fork it. This is functionality of RCloud AT&T .

**Comment by tkang007** [comment]
Hi there, 

This could be a simple option to share notebooks without giving write permission to each users in the shared folders by allowing list, view and run notebooks at jupyterhub.

a. create a dir at local, nfs(efs) or cloud storage filesystem(s3fs).

b. giving the ownership of the dir to a admin user, set gid to the group of the admin user, and allow read permission with setfacl for all new subdir and files.

c. create a symlink for the dir at /etc/skel before creating users.

d. create os users.

e. git clone any notebook repo t the dir.

I am looking for simple way to prvide notebook html viewing url on the dir without using jupyterhub. 

nbviewer might  be an option with/without nbconvert or papermill.

Is ther any doc for this?

**Comment by kkmann** [comment]
+1 one for finer grained authentification support. think of an institution sharing analyses with sensitive data internally.

**Comment by francois-travais** [comment]
@willingc It looks like HubShare is a dead project, what happened? Was it too difficult? Lack of resources?

Also @minrk mentioned a long time ago that something was planned, was it HubShare? Something else?

**Comment by willingc** [comment]
@francois-travais Quiet for a while mostly due to priorities and resources. @Zsailer has recently put effort into moving the project forward https://github.com/jupyterhub/hubshare/pull/17

**Comment by Zsailer** [comment]
@jackblackCH 

> Is it still not possible to "hack" together on a notebook file, like on a google docs file?
 
This is on the horizon. We have a core Jupyter developer working on ["real time collaboration" (RTC) inside of Jupyterlab](https://github.com/jupyterlab/jupyterlab/issues/5382) (see the bottom of the issue for most recent discussion). After this stabilizes, we should be in a good position to bring RTC to JupyterHub deployments. 

@kkmann 
> +1 one for finer grained authentification support

This is also on our radar. There are a couple core Jupyter/JupyterHub devs exploring how to effectively implement fine grained authentication in JupyterHub. This will likely happen in the [Hubshare](https://github.com/jupyterhub/hubshare) repo; though, it's still in the research/exploration phase. 

@francois-travais 
>t looks like HubShare is a dead project, what happened? Was it too difficult? Lack of resources?

It's not a dead project, just stalled. As @willingc said, we are constantly balancing priorities and resources. We're planning to pick up development soon!

**Comment by hbarnard** [comment]
I have a Code Club in the UK and am beginning to experiment with the Littlelest Hub. I've hacked together a script that distributes material from a 'master' Jupyter user to all the student accounts, where the student doesn't already have a copy (in which case their copy is assumed to be 'live'). This, of course, doesn't do, real-time-share etc. but I also didn't want NFS, cloud integration etc. wanted something a bit stupider.

**Comment by kipstakkr** [comment]
Is there any progress on Hubshare or the priorities are still in place for something else? If that's the case (which is there for a long time now), is there a workaround/hack to share files between users in Jupyterhub?

**Comment by jhegedus42** [comment]
just a hack : how about a soft link into each user's directory ?

**Comment by davidedelvento** [comment]
@jhegedus42 this (like many other "hacks" or workarounds described in this issue, including "share by email") will only work for

- read only
- not real time

which is kind of useless for most people interested in this problem...

**Comment by jhegedus42** [comment]
actually.... i just figured out a way using this thing...

what is real time ? 

what is read only ?

I just tested it, it works.

sure, it can be fancy and whatnot... and it takes 3 years to complete it

or sit down make 3 simlinks and get it done in 3 minutes

just as functional :)

it's not even a hack

it's just a bit of an out of the box thinking ? in which people are not interested ?

loool

seriously... give it a try... it works ... 

if it is good enough for a startup which has no 3 years to wait for solving a 
problem like this in the "right way" :) then ... maybe it is interesting for people
who cannot wait 3 years ... coz they want to get this done yesterday... which I
did...

I leave it up to your imagination, how I did it, (just a hint : I used soft symlinks)

Unfortunately I cannot show the code (which is not even a code) because it
contains sensitive information.

Seriously. People. Don't overengineer things. Go and pet some cute little Siamese
kittens instead :). Or they can be Burmese too. Or partially Burmese/Siamese streetcat mix.
(I hope [Paapuli](https://www.reddit.com/r/cats/comments/68cipd/paapuli_from_koivukyl%C3%A4/) is doing well. ) :( and :) simultaniously ... :( :) 

On the side : Jupyterhub is very nice already. So, whoever wrote it, I am super 
respectful and thankful. But there is also a limit to how much overengineering one
can take, in a startup, especially.

**Comment by anki-code** [comment]
@minrk @willingc @jasongrout please take a look to https://github.com/metabase/metabase .

This tool have just 3 simple but kill features that make it very easy to use and profitable and this features can be implemented in stack of JupyterHub+JupyterLab (JHJL):

1. 5 minutes UI for authentication setup (email, gsuite, ldap) and UI for users and groups permissions. In JupyterHub (JH) stack instead of it we have configs and code which basically not ready to use and should be set by hands.

2. 5 minutes UI for making SQL query (read jupyter notebook), saving it and sending link to results to any people. Permissions for access are setting on step 1. In JHJL we haven't any way to send link just for reading the result of notebook. I think the MVP of this feature is just render the notebook in the `voila` as a JH service or something like that.

3. Dashboards. Dashboards. Dashboards. Special notebooks which can be formatted as dashboards. I've seen jupyter/dashboards but it works with glitches and unuseful for me.

My dream is the Metabase where I can use Jupyter instead of SQL. Thanks!

**Comment by willingc** [comment]
Hi @jhegedus42,

> But there is also a limit to how much overengineering one
can take, in a startup, especially.

As an FYI, I want to clarify to you and others that Project Jupyter is not a startup. We're a non-profit that has grown from roots in academia 

[About Page - Jupyter website](https://jupyter.org/about)
[Post about Project Jupyter winning the ACM Software System Award which includes the Project's history](https://blog.jupyter.org/jupyter-receives-the-acm-software-system-award-d433b0dfe3a2?source=collection_archive---------5-----------------------)

P.S. I'm glad you are enjoying and using JupyterHub

**Comment by willingc** [comment]
@anki-code Thanks for sharing the information as other users may find it helpful. :sunny:

**Comment by jhegedus42** [comment]
> P.S. I'm glad you are enjoying and using JupyterHub

Yeah, it's the future. Super cool. I am kinda happy that my hack worked pretty nicely for this sharing issue. Now JH has everything a small startup could ever need :). I just created symlinks and gave proper permissions and now sharing is as easy as it can be... just copy the stuff you want to share into the "my_stuff_that_i_want_to_share_with_everyone" directory, and there you go.

This works fine for small teams. Ideally people who collaborate on the same notebook they do that by using git.

Also, ideally, people do not edit other people's notebooks.

Jupyter is not meant for writing code. IDEs are meant for that.

I don't recommend writing more than 10-20 lines of code in Jupyter, certainly not more than 50.

So, bottom line, Jupyter is not for shared coding on a large codebase, since once the code grows beyond 50 lines, it is time to switch to git+IDEs.

So, to be honest. Collaboration for code which is less than 50 lines ... a simple hacky solution is more than enough. If one wants to collaborate on more than 50 lines of code, then it is better done using an IDE combined with git.

Trust me, I am a doctor. :) llllooool , ok, usually I use this as a pickup line :) and it never worked :) llloooll

**Comment by Ben-Epstein** [comment]
Is there an update on this issue? Would love the ability to share notebooks between users

**Comment by krinsman** [comment]
@Ben-Epstein Please excuse the self-promotion (if possible), but I am currently working for NERSC on something which would allow one to do this (using NBViewer as a JupyterHub-run service): https://github.com/krinsman/clonenotebooks

Unfortunately it's under active development and not very stable. Anyway, if you can create a GitHub  repo with the notebooks, users could share the notebooks from that. Or if there is a logical directory "local" to your JupyterHub deployment, you could store the notebook files to be shared there, and then any users could look at the notebooks in NBViewer, and if they decide they like a particular notebook, clone it to their user server.

If you're interested in looking into it (I would _love_ any feedback I could get from users), probably the best place to start would be the Docker image I made containing a bare-bones example deployment: 
https://github.com/krinsman/clonenotebooks/tree/master/Docker

**Comment by whitespaceninja** [comment]
@jhegedus42 actually has a good point. If you have a small enough batch of users, you can symlink folders between the users. Sample:
```
sudo ln -s /home/jupyter-username1-abcd4 /home/jupyter-username2-z1234/shared-folder
```

This will provide username1 with access to username2's "shared-folder". If you are sharing multiple folders between each other, I would recommend a naming convention like "shared-username". It's not scalable, but for < 10 users it's pretty great. You may also need to restart your personal server to see it. 

Note: You can find everyone's jupyter username with `less /etc/passwd`.

**Comment by rebeca-maia** [comment]
> Here is a mockup
> ![shareable-anon 1](https://user-images.githubusercontent.com/30668418/39888890-1404f3a6-544b-11e8-8679-d8d1c442e311.png)

how did you do that?

**Comment by jhegedus42** [comment]
powerpoint?
photoshop?
c++?
c--?
c+?

**Comment by rebeca-maia** [comment]
I mean how he implemented this sharing mechanism into jupyterlab.

**Comment by jhegedus42** [comment]
i think it was using the paintbrush in photoshop or what else?

**Comment by jhegedus42** [comment]
was he using the pen tool too ? coz the lines look pretty sharp... i think
you cannot paint so sharp lines with paintbrush alone, right?

what do u guys think?

just paintbrush alone?
or
paintbrush + pen tool too?

i start to wonder myself, about the implementation details

hmm...

**Comment by ronaldpetty** [comment]
@whitespaceninja nice suggestion.  I was able to share with a few peers.  I did notice that writing files changes the group to the user only group.  I have not found it just yet, is there a place to change how the process writes the results to be owned by a shared group?  That way others can edit as well.  I could do epoll, chown; but that seems like a hack.

**Comment by whitespaceninja** [comment]
@ronaldpetty I haven't figured out a way to allow other users to edit them. Currently it's read-only, which for my purposes works pretty well (I'm a technical user giving scripts to non-technical users). It's a feature, not a bug :)
I imagine a 'proper' fix would have to be done in the source code itself. Otherwise your suggestion for polling could work or anything more complicated than that just warrants implementing the feature as a proper feature in the original source.

**Comment by ronaldpetty** [comment]
@whitespaceninja you are correct, should be done as a proper feature.  I did get closer after considering @tkang007 suggested acl usage.  I have to dig up my notes but will drop them in here.  I have use Jupyter for a while but not Hub, so maybe once I get my feet under me could try to see about creating a more straight-forward solution.

**Comment by harrybiddle** [comment]
I've also been trying to get shared folders working. I've got quite far with this:

```bash
# create a new group
groupadd shared
usermod -G shared jupyter-user

# create shared directory
mkdir /opt/shared
chgrp shared /opt/shared  #  set new group owner
chmod 770 /opt/shared  # owner and group has rwx
chmod +s /opt/shared  # new directories inherit permissions
```

Then, I mount that folder in the user's home directory.

This works fine except that new files created in the web interface have `-rw-r--r--`, which misses the write permission for a group. Is this configurable? My only idea is to run a stupid cron job just to constantly fix the permissions...

**Comment by jaredwinick** [comment]
@harrybiddle I followed your procedure but then also used `setfacl` as described in [https://unix.stackexchange.com/questions/1314/how-to-set-default-file-permissions-for-all-folders-files-in-a-directory](https://unix.stackexchange.com/questions/1314/how-to-set-default-file-permissions-for-all-folders-files-in-a-directory) to successfully have the files created by JupyterLab in the shared directory be group writable.

**Comment by gutow** [comment]
As I was setting up a littlest jupyter hub for my class this coming Fall, I built some utilities that put some of these ideas together. Maybe this will help people manage shared directories for their hub setups. I have put things in [github.com/gutow/tljh_grp_utils](https://github.com/gutow/tljh_grp_utils). Suggestions comments and additions welcome.

**Comment by sachin235** [comment]
I have developed the following two plugins:

1) https://github.com/sachin235/JhubGitPush
This enables the users to commit and push notebooks to GitHub from JupyterHub.

2) https://github.com/sachin235/JhubBitbucketPush
This enables the users to commit and push notebooks to Bitbucket from JupyterHub.

I have elaborated the details in the Readme. Feel free to report any issue or enhancements that can be incorporated!

**Comment by manics** [comment]
I'm closing this as it's quite old and there's no concrete action point in JupyterHub, there may be other projects which have tried to implement this as a JupyterHub add-on. If you'd like to discuss this further please post on the [Jupyter Community Forum](https://discourse.jupyter.org/) where more of the community hang out. Thanks!
